### PR TITLE
Add release process for the `github-actions` package

### LINKED ISSUE #15: Add GitHub workflows to help the release process of the `github-actions` packages in this repo
## User story

As a developer, it would be good if the release process could be done easier and clearer.

#### To make the new release more convenient, some automatic jobs were planned to be added to help the process:

1. 📌 Prepare a release content when a specific branch is created onto the target revision.
   - Fetch the automatically generated release notes provided by GitHub API.
   - Prepend the changelog content to *CHANGELOG.md*.
   - Bump the `version` in *package.json* and *package-lock.json*.
2. 📌 Based on the specific branch, create a release PR after a release content is finished.
   - Get the version from *package.json* to generate the PR title.
   - Add the release instruction in the description of the release PR.
3. 📌 Create a new release on GitHub after a release PR is approved.
   - Get the version from *package.json* to composite the release tag.
   - Get the respective content from *CHANGELOG.md* as the release notes.

#### With these workflows, the release process would be like:

1. When a package is ready to release, create a release branch onto `trunk`. For example, `release/actions`.
2. Check if the new changelog content and updated version are correct in the release PR.
   - If something needs to be revised,  append the changes in this release PR.
3. Approve the release PR if it's all good.
   - Check if the corresponding release workflows are run successfully. For example, the workflow of creating a new release or updating version tags.
4. Merge the release PR.

### PR DESCRIPTION
**Pull request by eason9487**
### Changes proposed in this Pull Request:

This PR implements parts 2 and 3 in #15 for the `github-actions` package. **📌 Based on the specific branch, create a release PR after a release content is finished.** and  **📌 Create a new release on GitHub after a release PR is approved.**

- Convert `getPackageSettings` util to `PackageTool`, and add a new method `getChangelogByVersion`.
- Add a new input to the update-version-tags action to support `workflow_run` event.
- Create a release PR for `github-actions` package after the release content is prepared.
- A new workflow for creating a new release for the `github-actions` package after approving a release PR.
- Make github-actions-release workflow support be triggered after github-actions-create-release workflow is complete.

#### 📹 Simulate the new release process

https://user-images.githubusercontent.com/17420811/175922055-21332a8b-8b38-4ea3-892d-67a62fa752fd.mp4

#### 📷 Dismiss the approval if the release workflow failed

image

### Detailed test instructions:

💡 Since these workflows need to merge this PR to the default branch before testing, I created a simulation to demonstrate the results in a forked repo. And a more complete process of new release can be found in the video above.

1. Triggered the ["Prepare New Release" workflow](https://github.com/eason9487/grow/actions/runs/2568496815) by creating a `release/actions` branch. And it created a release PR - https://github.com/eason9487/grow/pull/5
2. Trigger the ["Create Release" workflow](https://github.com/eason9487/grow/actions/runs/2568504434) after approving the release PR.
3. Trigger the ["Release" workflow](https://github.com/eason9487/grow/actions/runs/2568505679) to update version tags when a new release is published.

### CODE REVIEW

File: packages/js/github-actions/utils/package.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "type": "module"
+}
```
Comment (tomalec) [inline packages/js/github-actions/utils/package.json]: :grey_question: 
Why is that needed? Isn't the `packages/js/github-actions/package.json`'s `.type: "module"` enough?

File: .github/workflows/github-actions-release.yml [hunk 2]
```diff
@@ -6,33 +6,79 @@ on:
       - published
       - edited
 
+  workflow_run:
+    workflows:
+      - GitHub Actions - Create Release
```
Comment (tomalec) [inline .github/workflows/github-actions-release.yml:11]: :grey_question: 
Isn't the run duplicated? Meaning, doesn't `release.types.published` fire when the release is created by the  "Create Release" workflow?

File: .github/workflows/github-actions-create-release.yml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            }
+
+            const commentBody = `New release has been created: [${ release.name }](${ release.html_url })`;
+            await github.rest.issues.createComment( {
+              ...context.repo,
+              issue_number: context.payload.pull_request.number,
+              body: commentBody,
+            } );
+
+            fs.writeFileSync( '/tmp/release.json' , JSON.stringify( release ), 'utf8' );
```
Comment (tomalec) [inline .github/workflows/github-actions-create-release.yml]: :nail_care: Can we move those scripts to separate files? to get better syntax highlighting, lining, etc.

File: .github/workflows/github-actions-release.yml [hunk 4]
```diff
@@ -47,28 +93,39 @@ jobs:
         #
         # The second `git checkout HEAD^` is to prevent errors from the post jobs like "Prepare node".
         run: |
+          SOURCE_SHA=$(git rev-parse HEAD);
+          TAG_NAME="${{ github.event.release.tag_name }}"
+          if [ "$TAG_NAME" = '' ]; then
+            TAG_NAME="${{ fromJSON(needs.Setup.outputs.release).tag_name }}"
```
Comment (tomalec) [inline .github/workflows/github-actions-release.yml]: :grey_question: :nail_care: Maybe we could have a step that would resolve that condition once, and we would just reuse its output for checkout, and here?

**Review by tomalec — APPROVED** [review]
Oh my! I love the developer experience of this flow :smiley: 

I reviewed the code and workflow logs, LGTM.
Left a few cosmetic ideas and curiosity questions

File: packages/js/github-actions/utils/package.json (hunk 1 above)
Comment (eason9487) [reply packages/js/github-actions/utils/package.json]: Thanks for catching this. I didn't set package.json in the package root directory when experimenting in another repo and forgot here already has. Removed in https://github.com/woocommerce/grow/pull/22/commits/89644a30c92e3797fcfa4bb08ae319553e344592.

File: .github/workflows/github-actions-release.yml (hunk 2 above)
Comment (eason9487) [reply .github/workflows/github-actions-release.yml:11]: By GitHub's design, except for the `workflow_run` event, a workflow won't be triggered by events that are fired by a GitHub bot. So the `release.types.published` event won't fire this workflow when the release is created by the github-actions bot.

File: .github/workflows/github-actions-create-release.yml (hunk 3 above)
Comment (eason9487) [reply .github/workflows/github-actions-create-release.yml]: Sounds good! I separated them in https://github.com/woocommerce/grow/pull/22/commits/c9a3deda008dbf153892ceabe7a80171c5f58b01.

File: .github/workflows/github-actions-release.yml (hunk 4 above)
Comment (eason9487) [reply .github/workflows/github-actions-release.yml]: Thanks for the great idea! I adjusted this in https://github.com/woocommerce/grow/pull/22/commits/171be1d740de386a7ba9fefbb27872796c239600.

**Comment by eason9487** [comment]
@tomalec, thanks for the code review!

I made some adjustments according to your great suggestions! Since there were dozens of lines of code changes, could you help with another look?
- Simulation PR: https://github.com/eason9487/grow/pull/7
- Simulation workflows: [create PR](https://github.com/eason9487/grow/actions/runs/2607640028), [create release](https://github.com/eason9487/grow/actions/runs/2607647644), and [update version tags](https://github.com/eason9487/grow/actions/runs/2607648322).

File: .github/scripts/github-actions-create-release.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
+const fs = require( 'fs' );
+
+module.exports = async ( { github, context, workspace, outputJsonPath } ) => {
```
Comment (tomalec) [inline .github/scripts/github-actions-create-release.js]: :nail_care: :grey_question: Was there a reason to choose CJS over ESM? Personally, I like to stick to ESM wherever possible, as I consider CJS a legacy

**Review by tomalec — APPROVED** [review]
Thanks!
Re-reviewed the code and run logs, LGTM :)

File: .github/scripts/github-actions-create-release.js (hunk 5 above)
Comment (eason9487) [reply .github/scripts/github-actions-create-release.js]: > Was there a reason to choose CJS over ESM?

I didn't think about this and just follow the [example](https://github.com/actions/github-script#run-a-separate-file-with-an-async-function) to separate the script files. Changed in https://github.com/woocommerce/grow/pull/22/commits/e355fa39b4d4740b550758da36b2e982ec51ff3e and tested the changes by this simulation PR https://github.com/eason9487/grow/pull/10.

**Comment by eason9487** [comment]
Closing #15

**Comment by eason9487** [comment]
Close #15
